app/src/leader.c

```diff
@@ -0,0 +1,362 @@
+            release_leader_behavior(completed_sequence_candidates[i], k_uptime_get());
+        }
+    }
+    zmk_leader_deactivate();
+}
+
+static int position_state_changed_listener(const zmk_event_t *ev) {
+    struct zmk_position_state_changed *data = as_zmk_position_state_changed(ev);
+    if (data == NULL) {
+        return 0;
+    }
+
+    if (!leader_status && !data->state && !all_keys_released()) {
+        if (release_key_in_sequence(data->position)) {
+            return ZMK_EV_EVENT_HANDLED;
+        }
+        return 0;
+    }
+
+    if (leader_status) {
+        if (data->state) { // keydown
+            leader_find_candidates(data->position, press_count);
+            LOG_DBG("leader cands: %d comp: %d", num_candidates, num_comp_candidates);
+            stop_timer();
+            current_sequence[press_count] = data->position;
+            leader_pressed_keys[press_count] = data;
+            press_count++;
+            for (int i = 0; i < num_comp_candidates; i++) {
+                struct leader_seq_cfg *seq = completed_sequence_candidates[i];
+                if (seq->immediate_trigger || (num_candidates == 1 && num_comp_candidates == 1)) {
+                    press_leader_behavior(seq, data->timestamp);
+                }
+            }
+        } else { // keyup
+            if (data->position == active_leader_position && !first_release) {
+                first_release = true;
+                return 0;
+            }
+            if (!is_in_current_sequence(data->position)) {
+                return 0;
+            }
+            if (num_candidates == 0) {
+                zmk_leader_deactivate();
+                return ZMK_EV_EVENT_HANDLED;
+            }
+
+            release_count++;
+            release_key_in_sequence(data->position);
+
+            for (int i = 0; i < num_comp_candidates; i++) {
+                struct leader_seq_cfg *seq = completed_sequence_candidates[i];
+                if (seq->is_pressed && all_keys_released()) {
+                    release_leader_behavior(seq, data->timestamp);
+                    num_comp_candidates--;
+                }
+                if (num_candidates == 1 && num_comp_candidates == 0) {
+                    zmk_leader_deactivate();
+                }
+            }
+            if (!timerless || num_comp_candidates < num_candidates) {
+                reset_timer(data->timestamp);
+            }
+        }
+        return ZMK_EV_EVENT_HANDLED;
+    }
+
+    return 0;
+}
+
+ZMK_LISTENER(leader, position_state_changed_listener);
+ZMK_SUBSCRIPTION(leader, zmk_position_state_changed);
+
+#define LEADER_INST(n)                                                                             \
+    static struct leader_seq_cfg sequence_config_##n = {                                           \
+        .virtual_key_position = ZMK_KEYMAP_LEN + __COUNTER__,                                      \
```

Comment:
This should be refactored to use the virtual key position work @joelspadin added for combos, etc.